addon/lib/StudyTelemetryCollector.jsm

```diff
@@ -21,15 +22,266 @@ const BASERESOURCE = "esper-pioneer-shield-study";
+      Services.obs.addObserver(gatherPromise.resolve, "gather-telemetry");
+      TelemetrySession.observe(null, "idle-daily", null);
+      gatherPromise.promise.then(() => {
+
+        console.log("Received gather-telemetry notification.");
+        Services.obs.removeObserver(gatherPromise.resolve, "gather-telemetry");
+
+        try {
+
+          this.collectAndSendTelemetry();
+
+        } catch (ex) {
+          // TODO: how are errors during study execution reported?
+          // this.studyUtils.telemetryError();
+          Components.utils.reportError(ex);
+        }
+
+
+      });
+
+    });
+
+  }
+
+  async collectAndSendTelemetry() {
+
+    const telemetryPingEnvelopeBasedAttributes = StudyTelemetryCollector.collectTelemetryPingEnvelopeBasedAttributes();
+    const telemetrySubsessionPayloadBasedAttributes = StudyTelemetryCollector.collectTelemetrySubsessionPayloadBasedAttributes();
+
+    console.log("telemetryPingEnvelopeBasedAttributes", telemetryPingEnvelopeBasedAttributes);
+    console.log("telemetrySubsessionPayloadBasedAttributes", telemetrySubsessionPayloadBasedAttributes);
+
+    StudyTelemetryCollector.collectPlacesDbBasedAttributes().then((placesDbBasedAttributes) => {
+
+      console.log("placesDbBasedAttributes", placesDbBasedAttributes);
+
+      const shieldPingAttributes = {
+        event: "telemetry-payload",
+        ...telemetryPingEnvelopeBasedAttributes,
+        ...telemetrySubsessionPayloadBasedAttributes,
+        ...placesDbBasedAttributes,
+      };
+
+      const shieldPingPayload = StudyTelemetryCollector.createShieldPingPayload(shieldPingAttributes);
+
+      console.log("shieldPingPayload", shieldPingPayload);
+
+      this.studyUtils.telemetry(shieldPingPayload);
+
+    });
+
+  }
+
+  // TODO: @glind: move to shield study utils?
+  static createShieldPingPayload(shieldPingAttributes) {
+
+    const shieldPingPayload = {};
+
+    // shield ping attributes must be strings
+    for (const attribute in shieldPingAttributes) {
+      let attributeValue = shieldPingAttributes[attribute];
+      if (typeof attributeValue === "undefined") {
+        attributeValue = "null";
+      }
+      if (typeof attributeValue === "object") {
+        attributeValue = JSON.stringify(attributeValue);
+      }
+      if (typeof attributeValue !== "string") {
+        attributeValue = String(attributeValue);
+      }
+      shieldPingPayload[attribute] = attributeValue;
+    }
+
+    return shieldPingPayload;
+
+  }
+
+  /**
+   * These attributes are already sent as part of the telemetry ping envelope
+   * @returns {{}}
+   */
+  static collectTelemetryPingEnvelopeBasedAttributes() {
+
+    const telemetrySessionCurrentPingData = TelemetryController.getCurrentPingData();
+    console.log("telemetrySessionCurrentPingData", telemetrySessionCurrentPingData);
+
+    const environment = telemetrySessionCurrentPingData.environment;
+    const application = telemetrySessionCurrentPingData.application;
```

Comment:
You could get the same fields that you're using `application` for from the [environment](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/environment.html)

Reply:
Done, thanks